File: 2024/R/d01.R
parse_file <- function(f = "../inputs/input_1.txt") {
  read.csv(f, sep = "", header = FALSE, col.names = c("x", "y"))
}

part1 <- function(d) {
  sum(abs(sort(d$x) - sort(d$y)))
}

part2 <- function(d) {
  sum(unlist(lapply(d$x, function(x) sum(d$y[d$y == x]))))
}

d <- parse_file()
c(part1(d), part2(d))

File: 2024/Python/d01.py
from collections import Counter

def read(filename):
    x = y = list()
    with open(filename, "r") as file:
        x, y = zip(*[map(int, line.split()) for line in file])
    return(sorted(x), sorted(y))

def part1(x, y):
    return sum(abs(xi - yi) for xi, yi in zip(x, y))

def part2(x, y):
    ycount = Counter(y)
    return sum(item * ycount[item] for item in x)    

x, y = read("../inputs/input_1.txt")
print(part1(x, y))
print(part2(x, y))
